scylla/src/transport/session_builder.rs

```diff
@@ -260,6 +261,28 @@ impl SessionBuilder {
         self
     }
 
+    /// Set the fallback balancing policy
+    /// The default is Round-robin.
+    ///
+    /// # Example
+    /// ```
+    /// # use scylla::{Session, SessionBuilder};
+    /// # use scylla::transport::load_balancing::RoundRobinPolicy;
+    /// # use std::sync::Arc;
+    /// # async fn example() -> Result<(), Box<dyn std::error::Error>> {
+    /// let session: Session = SessionBuilder::new()
+    ///     .known_node("127.0.0.1:9042")
+    ///     .fallback(Arc::new(RoundRobinPolicy::new()))
+    ///     .build()
+    ///     .await?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn fallback(mut self, policy: Arc<dyn FallbackPolicy>) -> Self {
```

Comment:
The name `fallback` is way too generic for the `Session` class. But first and foremost, I generally think that a fallback shouldn't be added to a session, but instead directly to a policy, if it needs one. Specifically, does it make any sense to provide fallback for round robin or dc-aware round robin, if they already go over all nodes without any filtering? For now it looks like fallback would only be needed for:
1. TokenAware policy, where we already have a child policy and even a fallback mechanism used for different purpose (when the token is not present)
2. Custom policies, where users could define their own fallbacks as needed

I think that we don't really need this trait at all, and instead we should just make something like you mentioned in the description of #161:
```
A possible fix for this problem is to request the plan for a failed query again, but with token information erased. This would allow contacting nodes other than replicas.
```
Or, more specifically, instead of calling this line (which is the likely cause of our trouble):
```rust
self.child_policy.apply_child_policy(replicas)
```
we should go more for:
```rust
self.child_policy.apply_child_policy(replicas)
    .chain(self.child_policy.apply_child_policy(nonreplicas))
```
, which would naturally provide all known nodes to the user, in the desired order

Reply:
There is something i don't like in this approach. It is difficult to defer fallback plan calculation due to problems with lifetimes.

In current version fallback plan is calculated only when needed (after primary plan exhaustion). When trying to do a similar thing inside a `LoadBalancingPolicy::plan()` method I discovered that fallback plan calculation (`self.aply_child_policy()` or `self.plan()`) requires having a `&self` reference. Due to `LoadBalancingPolicy::plan()` signature being `fn plan<'a>(&self, statement: &Statement, cluster: &'a ClusterData) -> Plan<'a>`, returned plan cannot be dependent of `&self`, preventing deferred fallback calculation.

One solution to this problem is to ignore it :)) and don't implement deferring (maybe performance gain deferring is so low that it's not worth doing). Another solution would be to modify lifetimes in the signature of `LoadBalancingPolicy::plan()` to allow returned plans to be `&self` dependent.

@psarna @piodul what are your thoughts on this?

Reply:
We actually discussed it a little - I'm absolutely fine with not deferring anything. Especially in Scylla, when node count is usually counted in tens at best, we can afford taking all nodes and calculating the plan greedily

Reply:
I'm OK with not deferring, too. However, it's a bit sad that we do it because this will lead to more Arcs being cloned on each request and that reduces scalability. Perhaps we should consider measuring the impact and optimizing it later.